### ISSUE TITLE
Make URL encoding certain

### ISSUE THREAD

**Issue by ilatypov** [sqlmapproject/sqlmap#4239]
**Describe the bug**
An ASP.NET application needs `__VIEWSTATE=/wEP..+..+..SrR` and `__EVENTVALIDATION=/wEd..+..+..OY=` parameters containing special characters `+`, `/`.  The application also needs parameters `ctl00$...name=value` containing the dollar sign `$` in their names and a space ` ` in their values.

Submitting these (with proper protection against shell expansion) using `--data "NAME1=VALUE1&NAME2=VALUE2..."` shows unexpected responses.  (It's possible that the trace option `-t FILE` could help me in telling the wrong encoding but I don't know if I get the exact request body in its output).

**To Reproduce**
1. Use the above parameters.
2. See the unexpected response from the application.

**Expected behavior**
The application should give the same response as with curl's `--data-urlencode "NAME1=VALUE1" --data-urlencode "NAME2=VALUE2" ...`.

**Running environment:**
 - sqlmap version: 1.4.6#stable.
 - Installation method: homebrew.
 - Operating system: MacOS 10.15.5
 - Python version: 2.7.16.

As a work-around in my shell wrapper, I encoded every name and value with Python2's `urllib.quote(..., '')` (note the second argument that gets forward slashes `/` encoded) and submitted the concatenation of that with the ampersand character `&` as the `--data` parameter, then added the `--skip-urlencode`.  Due to another heuristics in sqlmap, I could not force the `Content-Type` header to `application/x-www-form-urlencoded` using the `-H` or `--headers` options, and after looking at the source code I came up with a dirty trick of adding `&a[]=b&a[]=b` to the `--data` value.  To avoid the "array" heuristics in that dirty trick, I removed it and `or conf.skipUrlEncode` in connect.py.
```diff
$ diff -u sqlmap/1.4.6/libexec/lib/request/connect.py{.orig,}
--- sqlmap/1.4.6/libexec/lib/request/connect.py.orig     2020-06-24 22:48:08.000000000 -0400
+++ sqlmap/1.4.6/libexec/lib/request/connect.py  2020-06-24 22:44:14.000000000 -0400
@@ -902,7 +902,7 @@
             headers = OrderedDict(conf.httpHeaders)
             contentType = max(headers[_] if _.upper() == HTTP_HEADER.CONTENT_TYPE.upper() else "" for _ in headers) or None

-            if (kb.postHint or conf.skipUrlEncode) and postUrlEncode:
+            if (kb.postHint) and postUrlEncode:
                 postUrlEncode = False
                 conf.httpHeaders = [_ for _ in conf.httpHeaders if _[1] != contentType]
                 contentType = POST_HINT_CONTENT_TYPES.get(kb.postHint, PLAIN_TEXT_CONTENT_TYPE)
```

**Comment by stamparm** [comment]
1) I am not sure that we are on the same page here. Please write concisely the command(s) you used, results got and results expected.

2) Example run with all these parameters (with latest revision - where dollar sign is treated "specially" inside the parameter names):

`python sqlmap.py -u 'http://debiandev/sqlmap/mysql/iis/get_int_60.asp' --data 'id=1&ctl00$name=value&__VIEWSTATE=/wEPAB+CDE+FGHSrR&__EVENTVALIDATION=/wEdABC+DEF+GHIOY='`

results with:

![1](https://user-images.githubusercontent.com/921555/85531952-89bf4600-b60f-11ea-808f-0b6b8dc42df4.png)

<- I don't see a problem of URL escaping of that `/` sign inside the `__VIEWSTATE` and `__EVENTVALIDATION`

3) In case of parameters containing `$` they should explicitly be enclosed with SINGLE quotes, not DOUBLE quotes because of the shell expansion of environment variables

**Comment by ilatypov** [comment]
Your screenshot show the very problem.  The plus characters would decode to spaces.  I see the same in the -t traffic file.

```
__EVENTVALIDATION=%2FwEdAJkBsrMKNHdbQsHoJE+KsVpf+Q1HlEW0iL7p%2FRzxa1sMNVfAIFqwG9UI6NJjDBjes3gRQdOEF4kmjdXeCeQTDmf4Khp3RLh1YW+Vd2SeBmdAmUFyYT65XFxgV6x3+GP+TdiwBZdQu26qLCzIfBaCckz7EZFvRFaoKO3Ppl7akYRzEAjhxIka0fHztabQiV84g5huA+a6fRAuIldDoZTqA90xZ%2F6FeLj91q6WNyfRcJT3LAh8lxK%2FYp+pZ3zug6jvkvBwuzH%2FXMpXIiZnOVYyEgLoziNW4Vs33tpdiZGTxhVw1+gvFKGsl5RkHXjxdPIm2nlnVWoJMNGAkS7%2F%2F%2FPOWD4Fjqe2w3XhDpqfhCJny%2F5esvXnvvdGw4nssId9l6bC2gS058N%2F+fKIDXvmoGAXZtQLOZBXbkX5n6chvl4s5XZp+AU0tdMGwbMfW5EbDDQ5xTEzLZfQ2kr9Ip4UVAy9EkTpsVEYxI5g+KU9er9qxs[...]
```

**Comment by stamparm** [comment]
So, you want those to be url-encoded. Sorry, but that is not clear from your original text.

**Comment by stamparm** [comment]
p.s. problem is that ASP(.NET) explicitly treats and prefers the usage of `+` as a whitespace, while other platforms use the regular `%20`. This behavior found in sqlmap is the RESULT of user requests where they insisted for the `+` to leave non-explicitly url-encoded (thus preventing double-encoding in majority of ASP(.NET) cases)

**Comment by stamparm** [comment]
With latest revision, your problem should be gone. Please don't force me to implement one more option which would be used by two users. If another similar Issue emerges I promise I'll do it

**Comment by ilatypov** [comment]
Thanks for the change, I hope it works (sorry I did not test it).  I did not mean to file this issue as a demand to fix it right away or as a demand to fix it at all.  I only thought of the issue as registering a mismatch for anyone (including myself) to consider or refute.

From my little corner, it's not the presence of special characters in the payload that demands their encoding.  Thinking the other way around, from when the decoding is applied, may clear this up.   This project's users requesting changes in the behaviour may indicate an unregistered issue that has a valid reason.   I apologize if am stepping on their heels without even knowing that.

The current misunderstanding around the plus sign, the space character, the `%20` and the `%2B` encoding highlights this.  One way to approach this is by hammering stuff until it works.  But I thought that web communications had these mentions of the above characters:
* The plus sign `+` encoding in the query string parameters of the request will be decoded by server-side URL decoders into a space to honour the founders, https://www.w3.org/History/19921103-hypertext/hypertext/WWW/Addressing/Search.html
* The `%20` encoding is a regular URL UTF-8 encoding for the space character invented later.
* The `%2B` encoding allows to pass the plus sign `+` as is.

As for the view state parameters containing the plus sign, I do observe them in the HTML responses.  Because they are `input` tag `value` attributes not `href` attributes, browsers are supposed (required) to read the attributes literally without URL-decoding them.  Then they would encode the plus signs with `%2B` so that the server-side query string or www-form-urlencode decoder decodes it back to `+`.

I apologize as I tip the scale into having the owner of the project prove that he is not guilty.  I only meant to start a healthy discussion.

**Comment by stamparm** [comment]
I believe that this concerns of yours were addressed with previous commit (74603c5). Commit is generic as it could be. Also, that `+` handling is something that is IMO problem in case of MsSQL. I could be a jerk and enforce the user to explicitly encode the data before being handed to the sqlmap (e.g. `+` to `%2b` in your case), BUT wanted to please everybody.

Anyhow, I believe that there is no need for further discussion here if the last commit suits your needs

**Comment by ilatypov** [comment]
No, the change did another heuristic based on the input parameter name.  (I appreciate the greater automation AND correct encoding done with the `--form` idea).
